crates/plugins/src/lookup.rs

```diff
@@ -41,37 +72,74 @@ impl PluginLookup {
             .map_err(|e| {
                 Error::ConnectionFailed(ConnectionFailedError::new(url.to_string(), e.to_string()))
             })?;
+
+        self.resolve_manifest_exact_from_good_repo(plugins_dir)
+    }
+
+    // This is split from resolve_manifest_exact because it may recurse (once) and that makes
+    // Rust async sad. So we move the potential recursion to a sync helper.
+    #[allow(clippy::let_and_return)]
+    pub fn resolve_manifest_exact_from_good_repo(
+        &self,
+        plugins_dir: &Path,
+    ) -> PluginLookupResult<PluginManifest> {
         let expected_path = spin_plugins_repo_manifest_path(&self.name, &self.version, plugins_dir);
-        let file = File::open(&expected_path).map_err(|e| {
-            Error::NotFound(NotFoundError::new(
-                Some(self.name.clone()),
-                expected_path.display().to_string(),
-                e.to_string(),
-            ))
-        })?;
-        let manifest: PluginManifest = serde_json::from_reader(file).map_err(|e| {
-            Error::InvalidManifest(InvalidManifestError::new(
+
+        let not_found = |e: std::io::Error| {
+            Err(Error::NotFound(NotFoundError::new(
                 Some(self.name.clone()),
                 expected_path.display().to_string(),
                 e.to_string(),
-            ))
-        })?;
-        Ok(manifest)
+            )))
+        };
+
+        let manifest = match File::open(&expected_path) {
+            Ok(file) => serde_json::from_reader(file).map_err(|e| {
+                Error::InvalidManifest(InvalidManifestError::new(
+                    Some(self.name.clone()),
+                    expected_path.display().to_string(),
+                    e.to_string(),
+                ))
+            }),
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound && self.version.is_some() => {
+                // If a user has asked for a version by number, and the path doesn't exist,
+                // it _might_ be because it's the latest version. This checks for that case.
+                let latest = Self::new(&self.name, None);
+                if let Ok(manifest) = latest.resolve_manifest_exact_from_good_repo(plugins_dir) {
```

Comment:
should we assert for the version number here (to ensure latest-version == version-requested)?

Reply:
Oops, yes, I think we should.  I did a lengthy series of refactorings on this mess and lost it along the way.  Thank you!